### PR TITLE
Responder Integration

### LINKED ISSUE #1142: Support for responder
Is there an easy way to integrate with responder web framework?

https://github.com/taoufik07/responder

### PR DESCRIPTION
**Pull request by clutchski**
Here's a first pass at a responder integration which addresses #1142.

Resolves #1142

### CODE REVIEW

**Comment by jd** [comment]
Could we make sure those _new_ files are blacked?

File: ddtrace/contrib/responder/patch.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    _set_patched(responder, False)
+
+    for obj, _, prop, _ in patches:
+        if '.' in prop:
+            attr, _, prop = prop.partition('.')
+            obj = getattr(obj, attr, object())
+        _u(obj, prop)
+
+
+async def _api_call(wrapped, instance, args, kwargs):
```
Comment (majorgreys) [inline ddtrace/contrib/responder/patch.py]: Pull these wrapper functions into a `wrappers.py`.

File: ddtrace/contrib/responder/patch.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+def _get_tracer(instance):
+    pin = ddtrace.Pin.get_from(instance)
+    if pin:
+        return pin.tracer
+    return ddtrace.tracer
+
+
+def _set_patched(obj, state):
+    setattr(obj, 'datadog_patch', state)
```
Comment (majorgreys) [inline ddtrace/contrib/responder/patch.py]: To follow the convention elsewhere:

```suggestion
    setattr(obj, '__datadog_patch', state)
```

Though in truth this is applied inconsistently across integrations.

File: ddtrace/contrib/responder/patch.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        traced_send = _trace_asgi_send_func(send, span)
+        span.set_tag(http.METHOD, scope.get('method'))
+
+        await wrapped(scope, receive, traced_send)
+
+        span.resource = scope.get('__dd_route', '404')
+
+
+def _api_template(wrapped, instance, args, kwargs):
+    # FIXME: needs tests
```
Comment (majorgreys) [inline ddtrace/contrib/responder/patch.py]: Let's add one if there's not on specifically for `API.template`

File: ddtrace/contrib/responder/patch.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            attr, _, prop = prop.partition('.')
+            obj = getattr(obj, attr, object())
+        _u(obj, prop)
+
+
+async def _api_call(wrapped, instance, args, kwargs):
+    tracer = _get_tracer(instance)
+    assert tracer
+
+    scope, receive, send = args  # FIXME: what if this is partially using kwargs?
```
Comment (majorgreys) [inline ddtrace/contrib/responder/patch.py]: We should be good since `API.__call__` only uses positional arguments:

https://github.com/taoufik07/responder/blob/master/responder/api.py#L361

```
async def __call__(self, scope, receive, send):
```

File: ddtrace/contrib/responder/patch.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        await wrapped(scope, receive, traced_send)
+
+        span.resource = scope.get('__dd_route', '404')
+
+
+def _api_template(wrapped, instance, args, kwargs):
+    # FIXME: needs tests
+    tracer = _get_tracer(instance)
+    with tracer.trace("responder.render_template") as span:
+        span.set_tag("template", args[0])  # FIXME args or kwargs?
```
Comment (majorgreys) [inline ddtrace/contrib/responder/patch.py]: `args[0]` is the `filename` positional argument. Being the underlying `jinja2` calls will get traced, adding this tag might be redundant metadata:

https://github.com/DataDog/dd-trace-py/blob/master/ddtrace/contrib/jinja2/patch.py#L52-L57

File: ddtrace/contrib/responder/patch.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        _w(mname, fname, func)
+
+
+def unpatch():
+    if not _is_patched(responder):
+        return
+    _set_patched(responder, False)
+
+    for obj, _, prop, _ in patches:
+        if '.' in prop:
```
Comment (majorgreys) [inline ddtrace/contrib/responder/patch.py]: Why is this necessary?

File: ddtrace/contrib/responder/patch.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+    scope, receive, send = args  # FIXME: what if this is partially using kwargs?
+
+    # check for propagated traced ids in the http heaers.
+    headers = {k.decode(): v for k, v in scope.get('headers', ())}
+    ctx = extract_context_from_http_headers(headers)
+    if ctx.trace_id:
+        tracer.context_provider.activate(ctx)
+
+    with tracer.trace("responder.request", service="responder") as span:
```
Comment (majorgreys) [inline ddtrace/contrib/responder/patch.py]: Import in `ddtrace.ext.SpanTypes` as we want to set the span type for these spans:

```suggestion
    with tracer.trace("responder.request", service="responder", span_type=SpanTypes.WEB) as span:
```

File: ddtrace/contrib/responder/patch.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    # FIXME: needs tests
+    tracer = _get_tracer(instance)
+    with tracer.trace("responder.render_template") as span:
+        span.set_tag("template", args[0])  # FIXME args or kwargs?
+        return wrapped(*args, **kwargs)
+
+
+def _api_template_string(wrapped, instance, args, kwargs):
+    tracer = _get_tracer(instance)
+    with tracer.trace("responder.render_template"):
```
Comment (majorgreys) [inline ddtrace/contrib/responder/patch.py]:
```suggestion
    with tracer.trace("responder.render_template", span_type=SpanTypes.TEMPLATE):
```

File: ddtrace/contrib/responder/patch.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        return wrapped(*args, **kwargs)
+
+
+async def _route_call(wrapped, instance, args, kwargs):
+    # patch the route on the scope to collect later.
+    scope = args[0]  # FIXME maybe kwarg?
```
Comment (majorgreys) [inline ddtrace/contrib/responder/patch.py]: Only positional arguments used:

https://github.com/taoufik07/responder/blob/master/responder/routes.py#L99

File: ddtrace/contrib/responder/patch.py (hunk 1 above)
Comment (clutchski) [reply ddtrace/contrib/responder/patch.py]: why? just convention?

File: ddtrace/contrib/responder/patch.py (hunk 6 above)
Comment (clutchski) [reply ddtrace/contrib/responder/patch.py]: I suspect it's possible its not there? copypasta from another integration

File: ddtrace/contrib/responder/patch.py (hunk 1 above)
Comment (majorgreys) [reply ddtrace/contrib/responder/patch.py]: a general convention that we don't enforce - i'm inclined to do separate the wrapped functions from the patching code, but might be overkill for simpler integrations.

File: ddtrace/contrib/responder/patch.py (hunk 6 above)
Comment (majorgreys) [reply ddtrace/contrib/responder/patch.py]: Right, I see it in the flask integration. I see the benefit of its use there being that we the functions patched differ across library versions.

With a fixed number of patched functions as here, a simpler solution might make more sense:

```
# in patch()
_w(responder.api.API, '__call__', _api_call)
_w(responder.api.API, 'template', _api_template)
_w(responder.api.API, 'template', _api_template_string)
_w(responder.routes.Route, '__call__', _route_call)

# in unpatch()
_u(responder.api.API, '__call__')
_u(responder.api.API, 'template')
_u(responder.api.API, 'template')
_u(responder.routes.Route, '__call__')
```

File: ddtrace/contrib/responder/patch.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    scope['__dd_route'] = getattr(instance, 'route')
+
+    await wrapped(*args, **kwargs)
+
+
+patches = [
+    (responder.api, 'responder.api', 'API.__call__', _api_call),
+    (responder.api, 'responder.api', 'API.template', _api_template),
+    (responder.api, 'responder.api', 'API.template_string', _api_template),
```
Comment (majorgreys) [inline ddtrace/contrib/responder/patch.py]: Not sure if this matters for the template functions you have 
```suggestion
    (responder.api, 'responder.api', 'API.template_string', _api_template_string),
```

File: tox.ini [hunk 11]
```diff
@@ -99,6 +99,8 @@ envlist =
     pyramid_contrib{,_autopatch}-{py27,py34,py35,py36,py37}-pyramid{17,18,19}-webtest
     redis_contrib-{py27,py34,py35,py36,py37}-redis{26,27,28,29,210,300}
     rediscluster_contrib-{py27,py34,py35,py36,py37}-rediscluster{135,136}-redis210
+# hi there
```
Comment (majorgreys) [inline tox.ini]:
```suggestion
```

File: ddtrace/contrib/responder/patch.py (hunk 1 above)
Comment (clutchski) [reply ddtrace/contrib/responder/patch.py]: i think i'll leave, relatively simple.

**Comment by Kyle-Verhoog** [comment]
This integration should be tweaked to follow along with the precedent set by https://github.com/DataDog/dd-trace-py/pull/1926 (reuse the ASGI middleware)
